### PR TITLE
Introducing checkpoints for scaffolder task action idempotency

### PR DESCRIPTION
**Pull request by acierto**
## Hey, I just made a Pull Request!

An implementation of checkpoints in scaffolder

Have tested checkpoints with using this template https://github.com/acierto/software-templates/blob/main/scaffolder-templates/docs-template/template.yaml There publish action called twice. Without checkpoints it will fail.

During the test I modified 'publish:github' action as follows:

```
const {
        _commitHash,
        _remoteUrl,
        _repoContentsUrl
      } = await ctx.checkpoint?.('repo.create', async(): Promise<{
        _commitHash: string,
        _remoteUrl: string,
        _repoContentsUrl: string
      }> => {
// basically almost all code was moved into callback here.
// I know that in a real scenario we will create multiple checkpoints, but to check that checkpoints work or not
// this simplification I believe is sufficient.
return {
          _commitHash: commitResult?.commitHash,
          _remoteUrl: remoteUrl,
          _repoContentsUrl: repoContentsUrl,
        }
      })!!;

      ctx.output('commitHash', _commitHash);
      ctx.output('remoteUrl', _remoteUrl);
      ctx.output('repoContentsUrl', _repoContentsUrl);
```

As a result I got a successful template run, where a second publish took 0 seconds and a record in a state column was:

```
{
  "state": {
    "checkpoints": {
      "v1.task.checkpoint.repo.create": {
        "status": "success",
        "value": {
          "_commitHash": "aae90a4863d93fb6bb1817ec22f9e46c00b47c55",
          "_remoteUrl": "https://github.com/acierto/testit.git",
          "_repoContentsUrl": "https://github.com/acierto/testit/blob/master"
        }
      }
    }
  }
}
```

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/scaffolder-backend/api-report.md [hunk 1]
```diff
@@ -519,11 +528,20 @@ export class TaskManager implements TaskContext {
   // (undocumented)
   emitLog(message: string, logMetadata?: JsonObject): Promise<void>;
   // (undocumented)
+  getCheckpoints?(): Promise<
```
Comment (acierto) [inline plugins/scaffolder-backend/api-report.md:531]: getTaskState

File: plugins/scaffolder-backend/api-report.md [hunk 2]
```diff
@@ -418,6 +425,8 @@ export class DatabaseTaskStore implements TaskStore {
     ids: string[];
   }>;
   // (undocumented)
+  saveCheckpoint(options: TaskStoreStateOptions): Promise<void>;
```
Comment (acierto) [inline plugins/scaffolder-backend/api-report.md:428]: optional

File: plugins/scaffolder-backend/api-report.md [hunk 3]
```diff
@@ -610,6 +634,12 @@ export type TaskStoreShutDownTaskOptions = {
   taskId: string;
 };
 
+// @public
+export type TaskStoreStateOptions = {
```
Comment (acierto) [inline plugins/scaffolder-backend/api-report.md:638]: inline it

File: plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts [hunk 4]
```diff
@@ -394,6 +397,28 @@ export class DatabaseTaskStore implements TaskStore {
...
 
+  async listCheckpoints({
+    taskId,
+  }: {
+    taskId: string;
+  }): Promise<{ state: TaskState }> {
+    const state = await this.db<RawDbTaskRow>('tasks')
+      .where({ id: taskId })
+      .select('state');
+    return { state: JSON.stringify(state) as unknown as TaskState };
```
Comment (acierto) [inline plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts:408]: should be parsed

File: plugins/scaffolder-backend/src/scaffolder/tasks/types.ts [hunk 5]
```diff
@@ -113,6 +113,16 @@ export type TaskStoreEmitOptions<TBody = JsonObject> = {
   body: TBody;
 };
 
+/**
+ * TaskStoreStateOptions
+ *
+ * @public
+ */
+export type TaskStoreStateOptions = {
```
Comment (acierto) [inline plugins/scaffolder-backend/src/scaffolder/tasks/types.ts:121]: inline

File: plugins/scaffolder-node/src/tasks/types.ts [hunk 6]
```diff
@@ -26,6 +26,38 @@ export type TaskSecrets = Record<string, string> & {
   backstageToken?: string;
 };
 
+/**
+ * The record passed to TaskBroker for updating a checkpoint.
+ * Parameters to store the result of the executed checkpoint
+ *
+ * @public
+ */
+export type CheckpointRecord =
```
Comment (acierto) [inline plugins/scaffolder-node/src/tasks/types.ts:35]: inline

File: plugins/scaffolder-node/src/tasks/types.ts [hunk 7]
```diff
@@ -26,6 +26,38 @@ export type TaskSecrets = Record<string, string> & {
...
+      status: 'failed';
+      reason: string;
+    };
+
+/**
+ * The state of all task's checkpoints
+ *
+ * @public
+ */
+export type TaskState = {
```
Comment (acierto) [inline plugins/scaffolder-node/src/tasks/types.ts:52]: inline

File: plugins/scaffolder-backend/api-report.md [hunk 8]
```diff
@@ -519,11 +563,40 @@ export class TaskManager implements TaskContext_2 {
...
+              status: 'failed';
+              reason: string;
+            }
+          | {
+              status: 'success';
+              value: JsonValue;
+            };
+      }
+    | undefined
+  >;
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/api-report.md:579]:
```suggestion
  getTaskState?(): Promise<
    | {
         state?: JsonObject
      }
    | undefined
  >;
```

File: plugins/scaffolder-backend/api-report.md [hunk 9]
```diff
@@ -573,6 +661,21 @@ export interface TaskStore {
...
+      [key: string]:
+        | {
+            status: 'failed';
+            reason: string;
+          }
+        | {
+            status: 'success';
+            value: JsonValue;
+          };
+    };
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/api-report.md:676]:
```suggestion
    state?: JsonObject
```

**Review by benjdlambert — COMMENTED** [review]
Some comments that are required

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.test.ts [hunk 10]
```diff
@@ -538,6 +559,59 @@ describe('NunjucksWorkflowRunner', () => {
...
+                  | {
+                      status: 'success';
+                      value: JsonValue;
+                    };
+              };
+            }
+          | undefined
+        > => {
+          return Promise.resolve({
+            state: {
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.test.ts:595]: Think this needs updating now right? Should be wrapped in `checkpoints` according to the latest changes in the BEP.

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts [hunk 11]
```diff
@@ -368,7 +368,9 @@ export class NunjucksWorkflowRunner implements WorkflowRunner {
             try {
               let prevValue: U | undefined;
               if (prevTaskState) {
-                const prevState = (prevTaskState.state as TaskState)?.[key];
+                const prevState = (
+                  prevTaskState.state?.checkpoints as TaskState
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts]: Think that `TaskState` in this case should be renamed to `CheckpointState` or something right now? Feels a little strange to call this `TaskState` now as it's not really a task

Comment (acierto) [reply plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts]: Haha, I just did that change few minutes ago :)

Comment (acierto) [reply plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts]: Do you prefer to have it as CheckpointState`? I named it `CheckpointsState`

Comment (benjdlambert) [reply plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts]: Well it depends. `CheckpointsState` would be the type of `state.checkpoints` but `CheckpointState` would be the type of `state.checkpoints.${id}` right? Probably don't need both, but I think it's better to use the narrower type of `CheckpointState` and just type `{[key: string}: CheckpointState}` for the `state.checkpoints`

**Review by benjdlambert — APPROVED** [review]
Nice :tada: let's get this in and keep iterating!

**Pull request merged**